File: lib/services/assessment_service.dart
import 'dart:convert';
import 'package:http/http.dart' as http;
import '../core/constants/api_config.dart';
import '../data/mock/mock_data_service.dart';
import '../models/student_model.dart';

class AssessmentService {
  static final List<AssessmentResult> _history = [];

  static List<AssessmentResult> get history => _history;

  Future<List<Question>> generateAssessmentQuestions({
    required String subject,
    required String topic,
    required String difficulty,
    required int questionCount,
  }) async {
    final url = Uri.parse('${ApiConfig.baseUrl}/api/assessment/generate');

    final payload = jsonEncode({
      'subject': subject,
      'topic': topic,
      'difficulty': difficulty,
      'question_count': questionCount,
      'student_context': {
        'student_name': MockDataService.currentStudent.name,
        'degree': MockDataService.currentStudent.degree,
      }
    });

    try {
      final response = await http
          .post(
            url,
            headers: {'Content-Type': 'application/json'},
            body: payload,
          )
          .timeout(ApiConfig.requestTimeout);

      if (response.statusCode == 200) {
        final data = jsonDecode(response.body);
        final questionsRaw = data['questions'] as List? ?? [];
        return questionsRaw.map((q) => _parseQuestion(q)).toList();
      }
    } catch (_) {}

    // Fallback question generation
    return MockDataService.cnQuestions.take(questionCount).toList();
  }

  Future<AssessmentResult> evaluateAssessment({
    required String assessmentId,
    required String subject,
    required String topic,
    required String difficulty,
    required List<Map<String, dynamic>> submissions,
    List<String>? adaptivePath,
  }) async {
    final url = Uri.parse('${ApiConfig.baseUrl}/api/assessment/evaluate');

    final payload = jsonEncode({
      'assessment_id': assessmentId,
      'subject': subject,
      'topic': topic,
      'difficulty': difficulty,
      'submissions': submissions,
      'adaptive_path': adaptivePath,
    });

    try {
      final response = await http
          .post(
            url,
            headers: {'Content-Type': 'application/json'},
            body: payload,
          )
          .timeout(ApiConfig.requestTimeout);

      if (response.statusCode == 200) {
        final data = jsonDecode(response.body);

        final res = AssessmentResult(
          score: data['score'] ?? 0,
          total: data['total'] ?? submissions.length,
          percentage: data['percentage'] ?? 0,
          strongAreas: List<String>.from(data['strong_areas'] ?? []),
          needsPractice: List<String>.from(data['needs_practice'] ?? []),
          aiRecommendation: data['ai_recommendation'] ?? 'Review weak topics.',
        );

        _history.add(res);
        return res;
      }
    } catch (_) {}

    // Fallback deterministic evaluation
    int score = 0;
    for (var sub in submissions) {
      if (sub['selected_key'] == sub['correct_key']) {
        score++;
      }
    }
    final pct = ((score / submissions.length) * 100).toInt();

    final fallbackRes = AssessmentResult(
      score: score,
      total: submissions.length,
      percentage: pct,
      strongAreas: const ['TCP basics', 'DNS'],
      needsPractice: const ['TCP connection termination'],
      aiRecommendation:
          'You may need more practice with TCP connection termination. Spend 10 minutes reviewing this topic before your next assessment.',
    );

    _history.add(fallbackRes);
    return fallbackRes;
  }

  Question _parseQuestion(Map<String, dynamic> json) {
    var optsRaw = json['options'] as List? ?? [];
    List<QuestionOption> options = optsRaw
        .map((o) => QuestionOption(key: o['key'] ?? '', text: o['text'] ?? ''))
        .toList();

    return Question(
      id: json['id'] ?? '',
      text: json['question'] ?? '',
      options: options,
      correctOptionKey: json['correct_key'] ?? 'A',
      explanation: json['explanation'] ?? '',
    );
  }
}

File: backend/app/services/assessment_service.py
from typing import Dict, List, Any, Optional
from ..models.assessment_models import (
    AssessmentGenerateRequest,
    AssessmentGenerateResponse,
    GeneratedQuestionModel,
    AssessmentEvaluateRequest,
    AssessmentEvaluateResponse,
)
from ..providers.assessment_provider import AssessmentProvider
from ..providers.gemini_assessment_provider import GeminiAssessmentProvider

class AssessmentService:
    def __init__(self, provider: Optional[AssessmentProvider] = None):
        self.provider = provider or GeminiAssessmentProvider()

    async def generate_assessment(self, req: AssessmentGenerateRequest) -> AssessmentGenerateResponse:
        raw_res = await self.provider.generate_questions(
            subject=req.subject,
            topic=req.topic,
            difficulty=req.difficulty,
            question_count=req.question_count,
            student_context=req.student_context
        )

        questions_raw = raw_res.get("questions", [])
        questions = [GeneratedQuestionModel(**q) for q in questions_raw]

        return AssessmentGenerateResponse(
            assessment_id=raw_res.get("assessment_id", "ass_101"),
            subject=raw_res.get("subject", req.subject),
            topic=raw_res.get("topic", req.topic),
            requested_difficulty=raw_res.get("requested_difficulty", req.difficulty),
            questions=questions,
            provider_used=raw_res.get("provider_used", self.provider.provider_name)
        )

    def evaluate_assessment(self, req: AssessmentEvaluateRequest) -> AssessmentEvaluateResponse:
        total = len(req.submissions)
        if total == 0:
            return AssessmentEvaluateResponse(
                assessment_id=req.assessment_id,
                score=0,
                total=0,
                percentage=0,
                strong_areas=[],
                needs_practice=[],
                ai_recommendation="Complete an assessment to get recommendations.",
            )

        correct_count = 0
        topic_scores: Dict[str, Dict[str, int]] = {}
        mistakes = []

        for sub in req.submissions:
            subtopic = sub.subtopic or sub.topic
            if subtopic not in topic_scores:
                topic_scores[subtopic] = {"correct": 0, "total": 0}

            topic_scores[subtopic]["total"] += 1

            is_correct = sub.selected_key.strip().upper() == sub.correct_key.strip().upper()
            if is_correct:
                correct_count += 1
                topic_scores[subtopic]["correct"] += 1
            else:
                mistakes.append({
                    "question_id": sub.question_id,
                    "question": sub.question_text,
                    "selected_key": sub.selected_key,
                    "correct_key": sub.correct_key,
                    "explanation": sub.explanation,
                    "subtopic": subtopic,
                })

        score_pct = int((correct_count / total) * 100)

        strong_areas = [t for t, data in topic_scores.items() if data["correct"] == data["total"]]
        needs_practice = [t for t, data in topic_scores.items() if data["correct"] < data["total"]]

        if not needs_practice and score_pct == 100:
            recommendation = f"Outstanding performance! You mastered all topics in {req.subject}."
        elif needs_practice:
            weak_str = ", ".join(needs_practice[:2])
            recommendation = f"You may need more practice with {weak_str}. Spend 10 minutes reviewing these topics before your next assessment."
        else:
            recommendation = f"Good effort! Review key concepts in {req.subject} to improve further."

        adaptive_summary = None
        if req.adaptive_path:
            path_str = " → ".join(req.adaptive_path)
            adaptive_summary = f"Adaptive Difficulty Path: {path_str}"

        # Trigger Sprint 5 Learning Profile Update
        try:
            from ..api.learning import learning_profile_service
            submissions_dict = [sub.model_dump() for sub in req.submissions]
            learning_profile_service.update_from_assessment_result(
                subject=req.subject,
                topic=req.topic,
                percentage=score_pct,
                submissions=submissions_dict,
                mistakes=mistakes
            )
        except Exception:
            pass  # Gracefully proceed if profile update encounters an issue

        return AssessmentEvaluateResponse(
            assessment_id=req.assessment_id,
            score=correct_count,
            total=total,
            percentage=score_pct,
            strong_areas=strong_areas if strong_areas else [f"{req.topic} basics"],
            needs_practice=needs_practice if needs_practice else [f"{req.topic} advanced concepts"],
            ai_recommendation=recommendation,
            adaptive_path_summary=adaptive_summary,
            mistakes=mistakes,
        )

